### PR TITLE
Support /team/foo deep linking

### PR DESCRIPTION
**Pull request by zapu**
This PR adds support for deep linking into `https://keybase.io/team/foo` pages which will result in navigating to a team tab of the specific team (assuming it exists).

Additionally there is a query parameter that's recognized: `applink`, which is optional and takes one of the following values right now: `add_or_invite`, `manage_settings`. `add_or_invite` will navigate to team builder of the team, and `manage_settings` will switch to Settings tab in navigated team page.

There are the following issues with this PR:
1) Navigating to team-builder does not stack team page properly (see my inline comment below).
2) Code for deep linking lives in at least two places and I don't understand why:
- `shared/actions/deeplinks.tsx` - this is where I added the team link support.
- `shared/actions/config/index.tsx` `routeToInitialScreen`
  - also `shared/actions/config/index.tsx` `maybeLoadAppLink` coming from `loadedSettings` action.

My PR seems to be working both when the app is in the background and when it's not launched (force-killed beforehand). 

Help @cjb @chrisnojima  ?

When this mystery is solved I'll improve the code in `urlToTeamDeepLink` by generalizing URL checks from `urlToUsername` (check protocol, check in-url username/password, check port etc.) and using them in all deep link parsing.

cc @keybase/y2ksquad @keybase/react-hackers 

EDIT:

Now also a `keybase://team-page/teamname/action` deep-link is supported. This gives us a support for this kind of link in desktop as well. Website PR makes it so `keybase.io/team/foo?applink=...` will try to redirect you to `keybase://` deep-link, and when Keybase is installed (and user accepts the redirect), it will show the appropriate tab / modal.

### CODE REVIEW

File: shared/actions/deeplinks.tsx [hunk 1]
```diff
@@ -85,6 +86,7 @@ const handleAppLink = (state: Container.TypedState, action: DeeplinksGen.LinkPay
   } else {
     // Normal deeplink
     const url = new URL(action.payload.link)
+    console.log('zzz Deep Link', action.payload.link, url, url.query, typeof url.query)
```
Comment (zapu) [inline shared/actions/deeplinks.tsx]: i need to ✂️ this

File: shared/actions/teams.tsx [hunk 2]
```diff
@@ -1481,6 +1504,8 @@ const teamsSaga = function*() {
 
   yield* Saga.chainAction2(TeamsGen.clearNavBadges, clearNavBadges)
 
+  yield* Saga.chainAction(TeamsGen.showTeamByName, showTeamByName)
```
Comment (zapu) [inline shared/actions/teams.tsx]: this action should probably be named differently

File: shared/actions/teams.tsx [hunk 3]
```diff
@@ -1402,6 +1402,29 @@ function* teamBuildingSaga() {
...
+  } catch (e) {
+    logger.info(`showTeamByName: team "${teamName} cannot be loaded: ${e.toString()}`)
+    return null
+  }
+  return [
+    RouteTreeGen.createNavigateAppend({
+      path: [
+        {props: {initialTab, teamID}, selected: 'team'},
+        // BUG: This doesn't actually go this route when addMembers is true, it
+        // just stacks team builder without stacking `team` screen first.
```
Comment (zapu) [inline shared/actions/teams.tsx]: bug here

File: shared/constants/config.tsx [hunk 4]
```diff
@@ -78,6 +71,33 @@ export const urlToUsername = (url: Url) => {
...
+  }
+
+  const teamName = match[1]
+  if (teamName.length < 2 || teamName.length > 255) {
+    return null
+  }
+
+  // `url.query` has a wrong type in @types/url-parse. It's a `string` in the
+  // code, but @types claim it's a {[k: string]: string | undefined}.
+  const queryString: string = url.query as any
```
Comment (buoyad) [inline shared/constants/config.tsx]: nit: you can do `as any as string`

**Comment by zapu** [comment]
@buoyad can you look again? some things have changed. thanks

File: shared/actions/teams.tsx [hunk 5]
```diff
@@ -1289,6 +1289,29 @@ function* teamBuildingSaga() {
   yield* Saga.chainAction(TeamsGen.addedToTeam, closeTeamBuilderOrSetError)
 }
 
+async function showTeamByName(action: TeamsGen.ShowTeamByNamePayload, logger: Saga.SagaLogger) {
+  const {teamName, initialTab, addMembers} = action.payload
+  let teamID: string
+  try {
+    teamID = await RPCTypes.teamsGetTeamIDRpcPromise({teamName})
+  } catch (e) {
+    logger.info(`showTeamByName: team "${teamName}" cannot be loaded: ${e.toString()}`)
```
Comment (buoyad) [inline shared/actions/teams.tsx]: probably worth `warn`

File: shared/actions/json/teams.json [hunk 6]
```diff
@@ -317,6 +317,11 @@
       "_description": "Set filtering for the subteams tab.",
       "filter": "string",
       "parentTeam?": "Types.TeamID"
+    },
+    "showTeamByName": {
+      "teamName": "string",
```
Comment (buoyad) [inline shared/actions/json/teams.json]: `teamname`

File: shared/actions/teams.tsx (hunk 5 above)
Comment (zapu) [reply shared/actions/teams.tsx]: This also happens when you get a link for a team you don't have an access to (or it doesn't exist). can also happen if you click on something from an e-mail, but you are logged in as someone else. So it's not necessarily an exceptional situation.

Comment (buoyad) [reply shared/actions/teams.tsx]: Ok, fair.

**Pull request merged**